File: problems/018-worker-generations/solutions/python/main.py
import sys

it = iter(sys.stdin.read().split())
n = int(next(it))
b = int(next(it))
gen = used = reject = 0
family = None
out = []
for _ in range(n):
    f = next(it)
    s = int(next(it))
    if s == 0:
        out.append("CACHE")
    elif s > 8 or s > b:
        reject += 1
        out.append("REJECT")
    else:
        if family != f or used + s > b:
            gen += 1
            family = f
            used = 0
        used += s
        out.append(f"WORKER {gen}")
out.append(f"SUMMARY {gen} {reject}")
print("\n".join(out))

File: problems/018-worker-generations/solutions/rust/main.rs
use std::io::{self, Read};
fn main() {
    let mut x = String::new();
    io::stdin().read_to_string(&mut x).unwrap();
    let mut it = x.split_whitespace();
    let n: usize = it.next().unwrap().parse().unwrap();
    let b: u64 = it.next().unwrap().parse().unwrap();
    let (mut generation, mut used, mut reject) = (0, 0u64, 0);
    let mut family = "";
    let mut out = String::new();
    for _ in 0..n {
        let f = it.next().unwrap();
        let s: u64 = it.next().unwrap().parse().unwrap();
        if s == 0 {
            out.push_str("CACHE\n")
        } else if s > 8 || s > b {
            reject += 1;
            out.push_str("REJECT\n")
        } else {
            if family != f || used + s > b {
                generation += 1;
                family = f;
                used = 0
            }
            used += s;
            out.push_str(&format!("WORKER {generation}\n"));
        }
    }
    out.push_str(&format!("SUMMARY {generation} {reject}\n"));
    print!("{out}");
}
